Makefile enhancements to optimize local workflows

huber-th opened the pull request:
## Advantage 360 Pro PR template

### What's changed:

Enhance the Makefile to enable firmware building exclusively for the left side.

### Why has this change been implemented:

To modify the keymap, flashing only the left side's firmware suffices.

This change facilitates the creation of firmware for the left side only, thereby expediting the build process when running locally via make.

### What (if any) actions must a user take after this change:

Usage: `make left` to build the left hand side firmware only

This update maintains existing functionality. Without specifying the left target, firmware for both halves will be generated. GitHub Actions also keeps building firmware for both halves on changes to the repo.

ReFil commented:
This is a really cool feature! I'd be open to reopening this and merging it into main

huber-th commented:
@ReFil while working on this I noticed that there are some deprecation warnings in the build logs. Most are for `label` which Zephyr deprecated in favor of using the devicetree node name as a node identifier.

https://github.com/zmkfirmware/zmk/pull/2028

I don't expect those to go away and therefore break the build but might be good to clean them up sooner than later.

ReFil commented:
Yeah that's on my list to sort, it'll be addressed as part of other planned upgrade work

huber-th commented:
I have another small update I made for my local process of building and cleaning by splitting the `make clean` into sub commands to allow for cleaning the firmware folder only without removing the docker image while maintaining current behaviour of `make clean`.

```
clean_firmware:
        rm -f firmware/*.uf2

clean_images:
        $(DOCKER) image rm zmk docker.io/zmkfirmware/zmk-build-arm:stable

clean: clean_firmware clean_images
```

Let me know if you find this useful and I can include it in this PR prior to merging.

ReFil commented:
That looks like another useful addition, yes please

huber-th commented:
Is there a benefit in adding the changed/updated `version.dtsi` file in a repo?

I wonder if it would be a good addition to add `$(shell git checkout config/version.dtsi)` as the final command of the build step in the `Makefile` to reduce noise by resetting that file after a build to avoid it showing as changed and being part of a commit or having to manually ensure it's not commited.

ReFil commented:
I think that's also a good idea. I couldn't think of a good way to ensure that file remains, but isn't changed as tricks like git assume unchanged dont carry over to other people's repos (and are kind of bad solutions anyway). I dont think there's anything wrong with it being changed but it's definitely not ideal to have it be changed and committed each time.

ReFil commented:
What OS have you been testing these changes on BTW?

ReFil commented on CHANGELOG.md:
```diff
@@ -4,6 +4,10 @@ Here's all notable changes and commits to both the configuration repo and the ba
 Many thanks to all those who have submitted issues and pull requests to make this firmware better!
 ## Config repo
 
+2/1/2024 - Add targets to makefile to allow for cleaning firmware and docker image separately while maintaining functionality of make clean to run both
```
Would you be able to link the PR in these changelog lines like the other changelog lines?

huber-th replied:
Will do. Would you want me to squash all changes into a single commit? I'm considering squashing and changing the commit message to better reflect all makefile changes as it is a bit more than just allowing left side build only by now.

ReFil replied:
Squashing into one would be appreciated

huber-th commented:
> What OS have you been testing these changes on BTW?

I'm working with macOS Sonoma 14.3

Unfortunately I don't have any Windows machines to verify the changes on myself.

ReFil commented:
Brilliant, I can test on Ubuntu and wsl2 but I don't have easy access to a macOS machine so between us we can cover all bases

huber-th commented:
Perfect. I will get the changes in later today and get it ready for you to test and verify.

huber-th commented:
Commits squashed, PR title updated and I updated the `README.md` with some additional info regarding building firmware locally on macOS.

huber-th committed:
Makefile enhancements to optimize local workflows

- Enhance the Makefile to enable firmware building exclusively for the left side.

To modify the keymap, flashing only the left side's firmware suffices.

This change facilitates the creation of firmware for the left side only,
thereby expediting the build process.

Usage: make left

This update maintains existing functionality. Without specifying the
left target, firmware for both halves will be generated.

- Add separate make targets to clean firmware and docker image

Add targets clean_firmware and clean_image independently while maintaing clean to run both

- Ensure files modified by the build process are reset at the end

The current build process dynamically generates the content of the version macro
in config/version.dtsi to enable users to print out the precise version and commit
the firmware was built on via a keybinding on the keymap.

This change ensurs that the changes to this macro file are reset at the end of the
build process to ensure it isn't shown as modified and added to keymap commit changes.

huber-th commented:
Small adjustment to the brew command to install `docker` and `colima` for macOS.

Ready for review and testing on other systems.

ReFil commented:
Everything works good on my ubuntu 20.04 system, i think it's ready to merge

huber-th commented:
> Brilliant, I can test on Ubuntu and wsl2 but I don't have easy access to a macOS machine so between us we can cover all bases

Feel free to tag me on future pull requests/changes if something needs to be tested on macOS specifically.
